core/translate.rs

```diff
@@ -313,86 +313,503 @@ fn translate_where(select: &Select, program: &mut ProgramBuilder) -> Result<Opti
+    select: &Select,
+    expr: &ast::Expr,
+) -> usize {
+    // This is the hackiest part of the code. We are finding the cursor that should be advanced to the next row
+    // when the condition is not met. This is done by introspecting the expression and finding the innermost cursor that is
+    // used in the expression. This is a very naive approach and will not work in all cases.
+    // Thankfully though it might be possible to just refine the logic contained here to make it work in all cases. Maybe.
+    let cursors = introspect_expression_for_cursors(program, select, expr).unwrap_or_default();
+    if cursors.is_empty() {
+        HARDCODED_CURSOR_LEFT_TABLE
+    } else {
+        *cursors.iter().max().unwrap()
+    }
+}
+
+fn evaluate_conditions(
     program: &mut ProgramBuilder,
     select: &Select,
-) -> Result<Option<BranchOffset>> {
-    // FIXME: clone()
-    // TODO: only supports INNER JOIN on a single condition atm, e.g. SELECT * FROM a JOIN b ON a.id = b.id, no AND/OR
+) -> Result<Option<QueryConstraint>> {
     let join_constraints = select
         .src_tables
         .iter()
         .map(|v| v.join_info.clone())
-        .filter_map(|v| v.map(|v| v.constraint.clone()))
-        .flatten()
+        .filter_map(|v| v.map(|v| (v.constraint.clone(), v.operator)))
         .collect::<Vec<_>>();
     // TODO: only supports one JOIN; -> add support for multiple JOINs, e.g. SELECT * FROM a JOIN b ON a.id = b.id JOIN c ON b.id = c.id
     if join_constraints.len() > 1 {
         anyhow::bail!("Parse error: multiple JOINs not supported");
     }
 
-    let maybe_join = join_constraints.first();
-
-    match (&select.where_clause, maybe_join) {
-        (Some(where_clause), Some(join)) => {
-            match join {
-                ast::JoinConstraint::On(expr) => {
-                    // Combine where clause and join condition
-                    let label = program.allocate_label();
-                    translate_condition_expr(program, select, where_clause, label, false)?;
-                    translate_condition_expr(program, select, expr, label, false)?;
-                    Ok(Some(label))
+    let join_maybe = join_constraints.first();
+
+    let constraint_maybe = match (select.where_clause.as_ref(), join_maybe) {
+        // JOIN tbl2 ON condition WHERE condition
+        (Some(where_clause), Some((Some(ast::JoinConstraint::On(expr)), op))) => {
+            let where_c = Where {
+                constraint_expr: where_clause.clone(),
+                no_match_jump_label: program.allocate_label(),
+                no_match_target_cursor: get_no_match_target_cursor(program, select, where_clause),
+            };
+            match op {
+                JoinOperator::TypedJoin { natural, join_type } => {
+                    if *natural {
+                        todo!("Natural join not supported");
+                    }
+                    // default to inner join when no join type is specified
```

Comment:
Looks like having all these conditions is too complex as there is duplicated code? I imagine you can have:
```
if where {
   do stuff with where
}
if join {
  do stuff with join
}
```
might be wrong tho

Reply:
less repetition now